### PR TITLE
MoE support using gmm kernels

### PR DESCRIPTION
**Pull request by blahBlahhhJ**

### CODE REVIEW

File: src/haliax/nn/linear.py [hunk 1]
```diff
@@ -124,3 +129,134 @@ def _out_first(self):
...
+    def __call__(self, inputs, group_sizes, *, key: Optional[PRNGKey] = None):
+        """
+        Args:
+            inputs (NamedArray): Input array    (Batch, Hidden)
+            group_sizes (NamedArray): MoE expert sizes (Experts)
+            key: Not used, but there for compat with other modules
+        """
+        del key
+
+        out_axes = inputs.axes[:-1] + (self.Out,)
```
Comment (dlwh) [inline src/haliax/nn/linear.py:191]:
```suggestion
        inputs = inputs.rearrange((...,  self.In)   # gmm requires axes at the end
        out_axes = hax.replace_axis(input.axes, self.In, self.Out)
       
```

File: src/haliax/nn/linear.py [hunk 2]
```diff
@@ -124,3 +129,134 @@ def _out_first(self):
...
+            In: AxisSpec: The input axis spec
+            Out: AxisSpec: The output axis spec
+            key: PRNGKeyArray: The PRNG key to use for initialization
+            use_bias: bool: Whether to use a bias term
+            out_first: bool: Whether to put output axes first in the weight matrix. out_first is how PyTorch does it.
+            dot_general: Callable: The dot_general function to use. Defaults to jax.lax.dot_general. For fp8 or int8
+            init_scale: float: The scale to use for initialization. We scale init by 1/sqrt(Input.size)*init_scale
+        """
+
+        # TODO: naive implementation only supports single axis mapping for now
```
Comment (dlwh) [inline src/haliax/nn/linear.py:168]: just update type sigs for now

File: src/haliax/nn/linear.py [hunk 3]
```diff
@@ -124,3 +129,134 @@ def _out_first(self):
...
+        # We do it this way because of scan layers
+        if isinstance(self.Out, hax.Axis):
+            return self.weight.axes[-1] != self.Out
+        else:
+            return self.weight.axes[-len(self.Out) :] != self.Out
+
+
+def _gmm(lhs, rhs, group_sizes, sharded=False):
+    if sharded:
+        gmm_fn = gmm_sharded
```
Comment (dlwh) [inline src/haliax/nn/linear.py:221]: can you document what this is for and when we should use it.

File: src/haliax/nn/linear.py [hunk 4]
```diff
@@ -124,3 +129,134 @@ def _out_first(self):
...
+    else:
+        gmm_fn = shard_map(
+            gmm_sharded,
+            mesh=hax.partitioning._get_mesh(),
+            in_specs=(
+                hax.partitioning.pspec_for_axis(lhs.axes),
+                hax.partitioning.pspec_for_axis(rhs.axes),
```
Comment (dlwh) [inline src/haliax/nn/linear.py:228]: this is wrong in the presence of FSDP (unless we're using a compute mapping) or TP? We want to force the all-gather right for FSDP.

To handle TP: up's `Out` axis is sharded so leaving it sharded is correct, ok, but on the way down, down's `In` axis is sharded and we want to do the AR on the way out. So I think we need to do a psum on the output if `In` is sharded.


For now I think we can just deshard the axis too, or assert

File: src/haliax/nn/linear.py [hunk 5]
```diff
@@ -124,3 +129,134 @@ def _out_first(self):
...
+        weight = hax.random.truncated_normal(key, joint_spec, -3, 3) * (init_scale / math.sqrt(input_size))
+        bias = hax.zeros(Out) if use_bias else None
+
+        return MoELinear(weight, bias, Experts, In, Out)
+
+    @named_call
+    def __call__(self, inputs, group_sizes, *, key: Optional[PRNGKey] = None):
+        """
+        Args:
+            inputs (NamedArray): Input array    (Batch, Hidden)
```
Comment (dlwh) [inline src/haliax/nn/linear.py:185]:
```suggestion
            inputs (NamedArray): Input array    (Batch, In)
```

File: src/haliax/nn/linear.py [hunk 6]
```diff
@@ -124,3 +129,134 @@ def _out_first(self):
...
+
+def _gmm(lhs, rhs, group_sizes, sharded=False):
+    if sharded:
+        gmm_fn = gmm_sharded
+    else:
+        gmm_fn = shard_map(
+            gmm_sharded,
+            mesh=hax.partitioning._get_mesh(),
+            in_specs=(
+                hax.partitioning.pspec_for_axis(lhs.axes),
```
Comment (dlwh) [inline src/haliax/nn/linear.py:227]: I think we want to force the all-gather. I think the right thing to do is to make a new axis mapping that deshards self.In, and deshard that axis mapping.

File: src/haliax/nn/linear.py (hunk 3 above)
Comment (blahBlahhhJ) [reply src/haliax/nn/linear.py:217]: This is for the case if someone wants to put the gmm function with some other code in the same shard_map. For example, I once tried putting all 3 gmm calls into one shard map. do you think I should keep it or not?

File: src/haliax/nn/linear.py (hunk 4 above)
Comment (blahBlahhhJ) [reply src/haliax/nn/linear.py:224]: For fsdp I don't think it's a problem. These are always called under compute mapping right?

**Review by dlwh — APPROVED** [review]
